Docker health

tjamet opened the pull request:
#### Short description of what this resolves:

Add support for docker based health checks

#### Changes proposed in this pull request:

This commit adds support for docker-managed health checks.
    
There are 2 ways of using it:
- either let docker manage the health check on its own and inspect the
   status thanks to the docker API
- or use the docker API to run a custom health check command in the
   container

dipak-pawar commented:
@tjamet [`DockerHealthAwaitStrategyTest` ](https://github.com/arquillian/arquillian-cube/pull/1004/files#diff-a161aeba78b437d8d0d71c4c15abcb3f) looks flaky. You can see it's failing for some time on CI. Build [Log](https://travis-ci.org/arquillian/arquillian-cube/jobs/347198942#L1784). Can you please look into it?

tjamet commented on docker/docker/src/test/java/org/arquillian/cube/docker/impl/await/DockerHealthAwaitStrategyTest.java:
```diff
@@ -0,0 +1,153 @@
+        testAwait(path, status, null);
+    }
+
+    private void testAwait(String imageId, boolean status, String[] command) {
```
done

tjamet commented on docker/docker/src/test/java/org/arquillian/cube/docker/impl/await/DockerHealthAwaitStrategyTest.java:
```diff
@@ -0,0 +1,153 @@
+    }
+
+    @Test
+    public void TestHealthCheckSuccess() {
```
done

tjamet commented on docker/docker/src/test/java/org/arquillian/cube/docker/impl/await/DockerHealthAwaitStrategyTest.java:
```diff
@@ -0,0 +1,153 @@
+        if (host == null || host.equals("")) {
+            host = "unix:///var/run/docker.sock";
+        }
+        dockerClient = DockerClientBuilder
```
I have used the default client constructor instead, watching the travis logs to see if that works in CI

tjamet commented:
@dipak-pawar Thank you for your feedback, I have addressed your points and added documentation, but went for letting the docker client to handle the socket path configuration instead

dipak-pawar commented on docker/docker/src/main/java/org/arquillian/cube/docker/impl/await/DockerHealthAwaitStrategy.java:
```diff
@@ -0,0 +1,107 @@
+        this.cube = cube;
+        this.client = dockerClientExecutor.getDockerClient();
+
+        if (params.getIterations() != null) {
```
How about moving this logic inside contrustor to one method and using it.

```java
    private int getIterations(Await params) {
        if (params.getIterations() != null) {
            return params.getIterations();
        } else {
            return DEFAULT_POLL_ITERATIONS;
        }
    }
```

dipak-pawar replied:
Can you check this last comment?

tjamet replied:
sorry missed it
done

dipak-pawar commented on docker/docker/src/test/java/org/arquillian/cube/docker/impl/await/DockerHealthAwaitStrategyTest.java:
```diff
@@ -0,0 +1,169 @@
+
+    public static DockerClient dockerClient;
+    public static String healthSuccessImageId;
+    public static String healthFailureImageId;
```
can't we make all(`dockerClient`, `healthSuccessImageId`, `healthFailureImageId` variables private here?

tjamet replied:
done

tjamet commented on docker/docker/src/test/java/org/arquillian/cube/docker/impl/await/DockerHealthAwaitStrategyTest.java:
```diff
@@ -0,0 +1,169 @@
+
+    @After
+    public void cleanContainers() {
+        for (int i = 0; i < containerIds.size(); i++) {
```
done

dipak-pawar commented on docker/docker/src/main/java/org/arquillian/cube/docker/impl/await/DockerHealthAwaitStrategy.java:
```diff
@@ -0,0 +1,113 @@
+
+    private static final int DEFAULT_POLL_ITERATIONS = 10;
+
+    private int pollIterations = DEFAULT_POLL_ITERATIONS;
```
This is not required as we are initializing it in constructor.

tjamet replied:
done

dipak-pawar commented on docker/docker/src/main/java/org/arquillian/cube/docker/impl/await/DockerHealthAwaitStrategy.java:
```diff
@@ -0,0 +1,113 @@
+
+public class DockerHealthAwaitStrategy extends SleepingAwaitStrategyBase {
+
+    protected static final Logger log = Logger.getLogger(DockerHealthAwaitStrategy.class.getName());
```
can't be private?

tjamet replied:
done

rpau commented:
@dipak-pawar when are you planning to release arquillian with this await strategy?

dipak-pawar commented:
@rpau We are planning to release new version of cube this week.
